Not working with styled-components v3.0.0

Broken as `styled-components` isn't shipping with `/lib` anymore (https://github.com/styled-components/styled-components/releases/tag/v3.0.1)

```
Cannot find module 'styled-components/lib/models/StyleSheet' from 'utils.js
```

--- a/README.md
+++ b/README.md
@@ -300,7 +300,7 @@ test('it works', () => {
 })
 ```
 
-The matcher supports a third `options` parameter which makes it possible to search for rules nested within an [At-rule](https://developer.mozilla.org/en/docs/Web/CSS/At-rule) ([media](https://developer.mozilla.org/en-US/docs/Web/CSS/@media) and [supports](https://developer.mozilla.org/en-US/docs/Web/CSS/@supports)) or to add modifiers to the class selector. This feature is supported in React only, and more options are coming soon.
+The matcher supports a third `options` parameter which makes it possible to search for rules nested within an [At-rule](https://developer.mozilla.org/en/docs/Web/CSS/At-rule) ([media](https://developer.mozilla.org/en-US/docs/Web/CSS/@media)) or to add modifiers to the class selector. This feature is supported in React only, and more options are coming soon.
 
 ```js
 const Button = styled.button`
@@ -314,7 +314,7 @@ const Button = styled.button`
 test('it works', () => {
   const tree = renderer.create(<Button />).toJSON()
   expect(tree).toHaveStyleRule('color', 'red', {
-    media: '(max-width: 640px)',
+    media: '(max-width:640px)',
     modifier: ':hover',
   })
 })

--- a/package.json
+++ b/package.json
@@ -45,13 +45,13 @@
     "react-dom": "^16.0.0",
     "react-native": "^0.49.3",
     "react-test-renderer": "^16.0.0",
-    "styled-components": "^2.1.1"
+    "styled-components": "^3.0.2"
   },
   "dependencies": {
     "css": "^2.2.1"
   },
   "peerDependencies": {
-    "styled-components": "^2.0.0"
+    "styled-components": ">=2.0.0"
   },
   "lint-staged": {
     "*.js": [

--- a/package.json
+++ b/package.json
@@ -51,7 +51,7 @@
     "css": "^2.2.1"
   },
   "peerDependencies": {
-    "styled-components": ">=2.0.0"
+    "styled-components": "^2.0.0 || ^3.0.2"
   },
   "lint-staged": {
     "*.js": [

--- a/src/utils.js
+++ b/src/utils.js
@@ -1,21 +1,43 @@
 const css = require('css')
-const { ServerStyleSheet } = require('styled-components')
-const StyleSheet = require('styled-components/lib/models/StyleSheet')
+const { ServerStyleSheet, isStyledComponent } = require('styled-components')
 
-const isOverV2 = () => Boolean(ServerStyleSheet)
+const isOverV3 = !!isStyledComponent
+const isOverV2 = () => !!ServerStyleSheet
+
+let StyleSheet
+
+/* eslint-disable */
+if (isOverV3) {
+  const secretInternals = require('styled-components')
+    .__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS
+
+  if (
+    secretInternals === undefined ||
+    secretInternals.StyleSheet === undefined
+  ) {
+    throw new Error(
+      'Could neither find styled-components secret internals nor styled-components/lib/models/StyleSheet.js'
+    )
+  } else {
+    StyleSheet = secretInternals.StyleSheet
+  }
+} else {
+  StyleSheet = require('styled-components/lib/models/StyleSheet').default
+}
+/* eslint-enable */
 
 const isServer = () => typeof document === 'undefined'
 
 const resetStyleSheet = () => {
   if (isOverV2()) {
-    StyleSheet.default.reset(isServer())
+    StyleSheet.reset(isServer())
   }
 }
 
 const getHTML = () =>
   isServer()
     ? new ServerStyleSheet().getStyleTags()
-    : StyleSheet.default.instance.toHTML()
+    : StyleSheet.instance.toHTML()
 
 const extract = regex => {
   let style = ''

--- a/test/toHaveStyleRule.spec.js
+++ b/test/toHaveStyleRule.spec.js
@@ -155,17 +155,11 @@ test('at rules', () => {
     @media (max-width: 640px) {
       color: green;
     }
-    @supports (display: flexbox) {
-      color: blue;
-    }
   `
 
   toHaveStyleRule(<Wrapper />, 'color', 'red')
   toHaveStyleRule(<Wrapper />, 'color', 'green', {
-    media: '(max-width: 640px)',
-  })
-  toHaveStyleRule(<Wrapper />, 'color', 'blue', {
-    supports: '(display: flexbox)',
+    media: '(max-width:640px)',
   })
 })
 
@@ -205,22 +199,13 @@ test('option combinations', () => {
         color: blue;
       }
     }
-    @supports (display: flexbox) {
-      &[href*='somelink.com'] {
-        color: green;
-      }
-    }
   `
 
   toHaveStyleRule(<Link />, 'color', 'black', {
     modifier: ':hover',
   })
   toHaveStyleRule(<Link />, 'color', 'blue', {
-    media: '(max-width: 640px)',
+    media: '(max-width:640px)',
     modifier: ':hover',
-  })
-  toHaveStyleRule(<Link />, 'color', 'green', {
-    supports: '(display: flexbox)',
-    modifier: "[href*='somelink.com']",
   })
 })

--- a/test/utils.spec.js
+++ b/test/utils.spec.js
@@ -1,22 +1,26 @@
 const { getHashes } = require('../src/utils')
 
-jest.mock('styled-components/lib/models/StyleSheet', () => ({
-  default: {
-    instance: {
-      toHTML() {
-        return `
-          <style data-styled-components="a">
-            /* sc-component-id: sc-1 */
-            .sc-1 {}
-            .a { color: red; }
-          </style>
-          <style data-styled-components="b c">
-            /* sc-component-id: sc-2 */
-            .sc-2 {}
-            .b { color: green; }
-            .c { color: blue; }
-          </style>
-        `
+jest.mock('styled-components', () => ({
+  ServerStyleSheet: true,
+  isStyledComponent: true,
+  __DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS: {
+    StyleSheet: {
+      instance: {
+        toHTML() {
+          return `
+            <style data-styled-components="a">
+              /* sc-component-id: sc-1 */
+              .sc-1 {}
+              .a { color: red; }
+            </style>
+            <style data-styled-components="b c">
+              /* sc-component-id: sc-2 */
+              .sc-2 {}
+              .b { color: green; }
+              .c { color: blue; }
+            </style>
+          `
+        },
       },
     },
   },

